nixos/modules/services/hardware/spacenavd.nix

```diff
@@ -0,0 +1,30 @@
+{ config, pkgs, lib, ... }:
+
+with lib;
+
+let
+  cfg = config.hardware.spacenavd;
+in
+{
+  options.hardware.spacenavd = {
+    enable = mkOption {
+      default = false;
+      example = true;
+      description = ''
+        Whether to enable spacenavd, a daemon which interprets the input from
+        3D mice in supported applications (Blender and OpenSCAD).
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    systemd.services.spacenavd = {
+      description = "Space Navigator 3D mice driver";
+      wantedBy = [ "multi-user.target" ];
+      partOf = [ "multi-user.target" ];
+      serviceConfig = {
+        ExecStart = "${pkgs.spacenavd}/bin/spacenavd -d";
```

Comment:
Last time I played with this, I opened 3 issues:

 - https://github.com/FreeSpacenav/spacenavd/issues/8
   This allows specifying a custom config file. We currently don't seem to configure anything?
 - https://github.com/FreeSpacenav/spacenavd/issues/9
  This introduces `-l syslog`, which we should probably set, to properly capture log output
 - https://github.com/FreeSpacenav/spacenavd/issues/10
  The `spnav.sock` socket currently is hardcoded, and I'm not sure about its permissions. Did you verify this works?

Reply:
I did manage to get spacenav working with this, but it's worth pointing out that spacenav joystick support is in Linux out of the box, and you can disable the annoying mouse movements with this:
```
  services.xserver.inputClassSections = [
    ''
      Identifier "3Dconnexion  SpacePilot PRO "
      MatchProduct "3Dconnexion  SpacePilot PRO "
      Driver "evdev"
      Option "Ignore" "on"
    '' ];

```
But blender doesn't support mouse navigation (openscad does).

- The config file is something which we could potentially capture with nix, but given it's got a gui configurator and it's very floating-point not easy to calculate you just have to try and see what works, we do definitely want to support that.
- Good idea
- I did verify, it works (when it's launched as root, I couldn't make it work non root though :( )

Reply:
Yeah, this could use some documentation, maybe the inputClassSection even a NixOS option.

Can you add some module documentation, like it's done with the nextcloud module?

> The config file is something which we could potentially capture with nix, but given it's got a gui configurator and it's very floating-point not easy to calculate you just have to try and see what works, we do definitely want to support that.

So you want to give this some cooking first?

> I did verify, it works (when it's launched as root, I couldn't make it work non root though :( )

GUI applications need to be launched as root? Or the spacenav daemon?

Reply:
The spacenav daemon needs to be root -- though you do need write permissions to the /etc/spnav.conf, which currently I think requires a chown.

Also, because my brain is a vegetable today, I forgot your first point, the OP did say they tried to get their patch accepted upstream, not sure if they have heard anything back from them.

I'll have a think about the config file, what I meant though is that it might not be applicable to be captured by nix, as the user might want to change it too often/faster than recompiling the system/with the gui rather than just picking values.

I'll try and remember to look at this in a couple of days when I am more functional, thanks for taking a look at this PR though!

Reply:
Having to no clear separation between user-wanted config and "state/calibration data" would indeed be a problem - let's iterate on that at a later point :-)